src/NServiceBus.Core/Persistence/InMemory/SagaPersister/InMemorySagaPersister.cs

```diff
@@ -3,203 +3,218 @@ namespace NServiceBus
     using System;
     using System.Collections.Concurrent;
     using System.Collections.Generic;
-    using System.Runtime.CompilerServices;
     using System.Threading.Tasks;
     using Extensibility;
-    using Persistence;
     using Sagas;
-
+    using Persistence;
+    
     class InMemorySagaPersister : ISagaPersister
     {
+        static readonly CorrelationId NoCorrelationId = new CorrelationId(typeof(object), "", new object());
+        readonly ConcurrentDictionary<Guid, Entry> sagas = new ConcurrentDictionary<Guid, Entry>();
+        readonly ConcurrentDictionary<CorrelationId, Guid> byCorrelationId = new ConcurrentDictionary<CorrelationId, Guid>();
+        static readonly JsonMessageSerializer serializer = new JsonMessageSerializer(null);
+
+        class Entry
+        {
+            public CorrelationId CorrelationId { get; }
+            readonly IContainSagaData data;
+
+            public Entry(IContainSagaData sagaData, CorrelationId correlationId)
+            {
+                CorrelationId = correlationId;
+                data = sagaData;
+            }
+
+            static IContainSagaData DeepClone(IContainSagaData source)
```

Comment:
This class could also implement `MemberWiseClone` or `IClonable`? Not sure it is helpful though